Lib/textwrap.py

```diff
@@ -429,42 +427,50 @@ def dedent(text):
 
     Entirely blank lines are normalized to a newline character.
     """
-    # Look for the longest leading string of spaces and tabs common to
-    # all lines.
-    margin = None
-    text = _whitespace_only_re.sub('', text)
-    indents = _leading_whitespace_re.findall(text)
-    for indent in indents:
-        if margin is None:
-            margin = indent
-
-        # Current line more deeply indented than previous winner:
-        # no change (previous winner is still on top).
-        elif indent.startswith(margin):
-            pass
-
-        # Current line consistent with and no deeper than previous winner:
-        # it's the new winner.
-        elif margin.startswith(indent):
-            margin = indent
-
-        # Find the largest common whitespace between current line and previous
-        # winner.
-        else:
-            for i, (x, y) in enumerate(zip(margin, indent)):
-                if x != y:
-                    margin = margin[:i]
-                    break
-
-    # sanity check (testing/debugging only)
-    if 0 and margin:
-        for line in text.split("\n"):
-            assert not line or line.startswith(margin), \
-                   "line = %r, margin = %r" % (line, margin)
+    # Fast paths for empty or simple text
+    if not text:
+        return text
 
-    if margin:
-        text = re.sub(r'(?m)^' + margin, '', text)
-    return text
+    if "\n" not in text:
+        return text  # Single line has no dedent
+
+    # Split text into lines, preserving line endings
+    lines = text.splitlines(keepends=True)
+
+    # Process in a single pass to find:
+    # 1. Leading whitespace of non-blank lines
+    # 2. Whether a line has zero leading whitespace (optimization)
+    non_blank_whites = []
+    has_zero_margin = False
+
+    for line in lines:
+        stripped = line.strip()
+        if stripped:  # Non-blank line
+            leading = line[:len(line) - len(line.lstrip())]
```

Comment:
```suggestion
        left_stripped = line.strip()
        if left_stripped:
            leading = line[:len(line) - len(left_stripped)]
```
If the `lstrip` is empty, then so is the `strip`